flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/ExprCodeGenerator.scala

```diff
@@ -679,13 +679,15 @@ class ExprCodeGenerator(ctx: CodeGeneratorContext, nullableInput: Boolean)
 
       // casting
       case CAST =>
+        val legacyCastEnabled = ctx.tableConfig
+          .get(ExecutionConfigOptions.TABLE_EXEC_LEGACY_CAST_BEHAVIOUR)
+          .isEnabled
         generateCast(
           ctx,
           operands.head,
-          resultType,
-          nullOnFailure = ctx.tableConfig
-            .get(ExecutionConfigOptions.TABLE_EXEC_LEGACY_CAST_BEHAVIOUR)
-            .isEnabled)
+          if (legacyCastEnabled) resultType.copy(true) else resultType,
```

Comment:
should we remove legacyCastEnabled for Flink 2?

Reply:
may not, the configuration ExecutionConfigOptions.TABLE_EXEC_LEGACY_CAST_BEHAVIOUR is not marked deprecated now

Reply:
Ok that makes sense. Shall we label ExecutionConfigOptions.TABLE_EXEC_LEGACY_CAST_BEHAVIOUR as deprecated in this fix, so people are dissuaded from using it?